src/secret.ts

```diff
@@ -0,0 +1,332 @@
+ *   const { username, password } = response.SecretValue!;
+ *
+ *   // secret is automatically serialized to JSON
+ *   await secret.putSecretValue({
+ *     SecretValue: {
+ *       username: "sam",
+ *       password: "password123"
+ *     }
+ *   });
+ * });
+ * ```
+ */
+export class JsonSecret<SecretValue = any> extends Secret<SecretValue> {
+  protected serialize(secretValue: SecretValue) {
+    return {
+      SecretString: JSON.stringify(secretValue),
+    };
+  }
+
+  protected async deserialize(
+    response: SerializedSecretValue
+  ): Promise<SecretValue> {
+    const data = await parse(response);
+    return JSON.parse(typeof data === "string" ? data : data.toString("utf8"));
+  }
+}
+
+/**
+ * Securely stores a string-encoded secret.
+ *
+ * ```ts
+ * const secret = new TextSecret(scope, "id");
+ *
+ * new Function(scope, "foo", async () => {
+ *   const text = await secret.getSecretValue();
+ *
+ *   await secret.putSecretValue({
+ *     SecretValue: "secret key 123"
+ *   });
+ * });
+ * ```
+ */
+export class TextSecret extends Secret<AWS.SecretsManager.SecretStringType> {
+  protected serialize(secretValue: string) {
+    return {
+      SecretString: secretValue,
+    };
+  }
+
+  protected async deserialize(
+    secretValue: SerializedSecretValue
+  ): Promise<string> {
+    const data = await parse(secretValue);
+    return typeof data === "string" ? data : data.toString("utf8");
+  }
+}
+
+/**
+ * Securely stores a binary-encoded secret.
+ *
+ * ```ts
+ * const secret = new TextSecret(scope, "id");
+ *
+ * new Function(scope, "foo", async () => {
+ *   const bytes = await secret.getSecretValue();
+ *
+ *   await secret.putSecretValue({
+ *     SecretValue: Buffer.from("secret key 123", "utf8")
+ *   });
+ * });
+ * ```
+ */
+export class BinarySecret extends Secret<Buffer> {
+  protected serialize(secretValue: Buffer): SerializedSecretValue {
+    return {
+      SecretBinary: secretValue,
+    };
+  }
+
+  protected async deserialize(
+    secretValue: SerializedSecretValue
+  ): Promise<Buffer> {
+    const data = await parse(secretValue);
+    return typeof data === "string" ? Buffer.from(data, "utf8") : data;
+  }
+}
+
+async function parse(
+  secretValue: SerializedSecretValue
+): Promise<string | Buffer> {
+  if (secretValue.SecretString) {
+    return secretValue.SecretString;
+  } else if (typeof secretValue.SecretBinary === "string") {
+    return secretValue.SecretBinary;
+  } else if (Buffer.isBuffer(secretValue.SecretBinary)) {
+    return secretValue.SecretBinary.toString("utf8");
+  } else if (secretValue.SecretBinary instanceof Uint8Array) {
+    return Buffer.from(secretValue.SecretBinary).toString("utf8");
+  } else if (secretValue.SecretBinary instanceof Blob) {
+    return Buffer.from(await secretValue.SecretBinary.arrayBuffer()).toString(
+      "utf8"
+    );
+  } else {
+    throw new Error(
+      `unable to parse SecretValue, expected a string or Buffer, but received `
+    );
```

Comment:
Synth error?

Reply:
It does not happen during synth though?

Reply:
Maybe we should rename SynthError to FunctionlessErr, or FunError lol